src/ripple/app/main/Application.cpp

```diff
@@ -1730,21 +1754,21 @@ ApplicationImp::setup()
     {
         using namespace boost::filesystem;
 
-        auto stateDb(
+        auto const stateDb(
             RPC::ShardArchiveHandler::getDownloadDirectory(*config_) /
             stateDBName);
 
         try
         {
+            // Create a ShardArchiveHandler to be initialized
+            // from the contents of the pre-existing state
+            // database.
             if (exists(stateDb) && is_regular_file(stateDb) &&
-                !RPC::ShardArchiveHandler::hasInstance())
+                !hasShardArchiveHandler())
```

Comment:
Can you describe the chain of events in the race condition?

Reply:
Sure.  There are two threads that are calling `getShardArchivehandler()` at about the same time.  One of those threads is in this code.

1. The code in this thread calls `hasShardArchiveHandler()` which returns `false`.  The other thread has not quite entered `getShardArchiveHandler()`.

2. The other thread enters `getShardArchiveHandler()`.

3. This thread enters `getShardArchiveHandler()` which, by this time, has a fully constructed `ShardArchiveHandler` despite `hasShardArchiveHandler()` having returned `false` just a few lines prior.

Does that sequence of events make sense?

Given this state of affairs, you have to make `getShardArchiveHandler()` robust to being called by multiple threads.  Additionally, if `getShardArchiveHandler()` _is_ robust to multiple threads, then you don't need to protect calls to `getShardArchiveHandler()` by checking `hasShardArchiveHandler()`.  Let `getShardArchiveHandler()` handle the situation for itself.

If an entry point is intended to be used in a multithreaded way, then putting additional (partially effective) guards around it is usually pointless.  It wastes compute cycles.  It leaves readers wondering if the entry point is really thread safe.

Reply:
Great, this is the scenario I thought you were implying.

I'm not sure its a race case though, since initialization will fail for one of the two threads. This is somewhat desired. Unit tests excluded, construction of the `ShardArchiveHandler` should be invoked only by either the `Application` at startup (due to preexisting state) or the relevant RPC handler as a result of an admin download request. In the scenario you described, this thread would attempt to initialize the `ShardArchiveHandler` which would fail and be logged. This is good since at this point in the code (during `Application` setup) it shouldn't be possible for another thread to attempt initialization. Alternatively, if this thread enters `getShardArchiveHandler` first, the other thread's attempt to initialize the handler will fail and be logged as well.

At this point in the code, it's imperative that `getShardArchiveHandler` constructs the handler, rather than returning a preexisting instance. So, how do you feel about removing the calls to `hasShardArchiveHandler` and `getShardArchiveHandler` and instead we:

1. Lock the `shardArchiveHandlerMutex_` within the current scope

2. Check that `shardArchiveHandler_ == nullptr`

3. Construct the `ShardArchiveHandler` if the aforementioned check was true; fail and log otherwise

Now this means that `getShardArchiveHandler` will only be called by external classes, which is fine. However, regarding another comment you've left, I still think its valuable to keep `hasShardArchiveHandler` for the sake of unit tests. It's a convenient way to confirm that the `ShardArchiveHandler` gets constructed automatically at startup.

Thoughts?

Reply:
Is the reason that it is imperative that a fresh `ShardArchiveHandler` be constructed at this point in the code because you want to be sure to call `InitFromDB()`?  You no longer have to worry about that here.  `init()` takes care of that for you.  If there's a pre-existing database and anybody calls `getShardArchiveHandler()` (from here or elsewhere), then `InitFromDB()` will be called.  Is there a different concern?

Reply:
That's not it - I realize that database initialization will still occur. It seems safer to enforce this because at this location in the code, it shouldn't be possible for the `ShardArchiveHandler` to have been constructed, meaning that if it's already present at this point, something unexpected happened which is cause for concern. Do you agree with keeping this expectation, and if so what do you think of the solution I proposed?

Reply:
Your current code does not raise such an alarm if the `ShardArchiveHandler` was constructed before you got to this point.  The way I read your current code this entire if condition is skipped if `hasShardArchiveHandler()` returns `true`.  Additionally, if `hasShardArchiveHandler()` returns `true` there are no writes to the log.  Your current code does not achieve what you say you want.

But beyond that, there are lots of places in the code where unexpected things may happen that would be cause for concern.  We can't guard them all.  In this particular case you've designed the initialization so that:

1. `getShardArchiveHandler()` will _always_ return either a `nullptr` or a fully initialized `ShardArchiveHandler`.  Hurrah!
2. If an appropriate shard archive handler database is found sitting on disk then the `ShardArchiveHandler` will initialize with it.  Otherwise the `ShardArchiveHandler` initializes empty.  Hurrah!

When the initialization scheme had a different design, then the concerns you are raising were important.  The initialization scheme has changed.  These concerns are way less important than they used to be.  To me it seems like you now have all of the important bases covered.  The thing you are worried about used to be important.  I don't think it is anymore.  This particular concern about this particular unexpected case doesn't strike me as being worth the effort.

But you also have to do what feels comfortable for you.  If you feel strongly, then code it the way you want and leave behind a paragraph comment justifying the approach.

Reply:
> Your current code does not raise such an alarm if the ShardArchiveHandler was constructed before you got to this point. The way I read your current code this entire if condition is skipped if hasShardArchiveHandler() returns true. Additionally, if hasShardArchiveHandler() returns true there are no writes to the log. Your current code does not achieve what you say you want.

Right, I'm planning on adding this safety check.

> When the initialization scheme had a different design, then the concerns you are raising were important. The initialization scheme has changed. These concerns are way less important than they used to be. To me it seems like you now have all of the important bases covered. The thing you are worried about used to be important. I don't think it is anymore. This particular concern about this particular unexpected case doesn't strike me as being worth the effort.

> But you also have to do what feels comfortable for you. If you feel strongly, then code it the way you want and leave behind a paragraph comment justifying the approach.

Understood, but I think this is a very simple check to add. Will add a comment.